linera-witty-macros/src/wit_export/caller_type_parameter.rs

```diff
@@ -0,0 +1,284 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Representation of a shared generic type parameter for the caller.
+
+use crate::util::{Specialization, Specializations};
+use proc_macro2::TokenStream;
+use proc_macro_error::abort;
+use quote::ToTokens;
+use std::collections::HashMap;
+use syn::{
+    punctuated::Punctuated, AngleBracketedGenericArguments, AssocType, GenericArgument, Generics,
+    Ident, PathArguments, PathSegment, PredicateType, Token, TraitBound, TraitBoundModifier, Type,
+    TypeParam, TypeParamBound, TypePath, WhereClause, WherePredicate,
+};
+
+/// Information on the  generic type parameter to use for the caller parameter, if present.
+#[derive(Clone, Copy, Debug)]
+pub enum CallerTypeParameter<'input> {
+    NotPresent,
+    WithoutUserData(&'input Ident),
+    WithUserData {
+        caller: &'input Ident,
+        user_data: &'input Type,
+    },
+}
```

Comment:
I'd be tempted to write
```rust
pub struct CallerTypeParameter<'input> {
    caller: &'input Ident,
    user_data: Option<&'input Type>,
}
```
and talk about `Option<CallerTypeParameter>` elsewhere.

Reply:
Done.